fuzzers/main.py

```diff
@@ -0,0 +1,842 @@
+        client.send_message({
+            'type':
+                'stream',
+            'to':
+                'pagoda/fuzzer/private',
+            'topic':
+                f'{branch}: artifact {artifact}',
+            'content':
+                f'''\
+# Fuzzer found new crash for branch *{branch}* {already_reported_msg}
+
+Full logs are available at {logs_url}.
+
+You can download the artifact by using the following command (all commands \
+are to be run from the root of `nearcore`):
+```
+gsutil cp {gcs_artifact} {artifact_dest}
+```
+
+Then, you can reproduce by running the following command:
+```
+cd {quoted_crate}
+RUSTC_BOOTSTRAP=1 cargo fuzz run {quoted_runner} {quoted_artifact}
+```
+
+Or minimize by running the following command:
+```
+cd {quoted_crate}
+RUSTC_BOOTSTRAP=1 cargo fuzz tmin {quoted_runner} {quoted_artifact}
+```
+
+Please edit the topic name to add more meaningful information once investigated. \
+Keeping the artifact hash in it can help if the same artifact gets detected as \
+crashing another branch.
+'''
+        })
+        focus_log_lines = '\n'.join(log_lines[:5]) + '\n[...]\n'
+        for line in log_lines[::-1]:
+            if line.startswith('```'):
+                # Censor the end of a spoiler block, not great but this is for human consumption
+                # anyway
+                line = ' ' + line
+            if len(focus_log_lines) + len(line) > 9000:
+                # Zulip limit is 10k, let's keep some safety buffer here
+                break
+            focus_log_lines = line + focus_log_lines
+        client.send_message({
+            'type': 'stream',
+            'to': 'pagoda/fuzzer/private',
+            'topic': f'{branch}: artifact {artifact}',
+            'content': f'```spoiler First and last few log lines\n{focus_log_lines}\n```',
+        })
+
+    def signal(self, sig: int) -> None:
+        """
+        Signal the fuzzer process (with all its process group as a fuzzer can spawn sub-fuzzers)
+        """
+
+        print(f'Sending signal {sig} to fuzzer {self.proc.pid}', file=sys.stderr)
+        os.killpg(os.getpgid(self.proc.pid), sig)
+
+
+# Actually this should also require a ['weight']: int bound but it seems like a mess to encode
+T = typing.TypeVar('T')  # pylint: disable=invalid-name
+
+
+def random_weighted(array: typing.List[T], num: int, name: str) -> typing.List[T]:
+    print(f'Picking {num} random {name} among {array}', file=sys.stderr)
+    untyped_array = typing.cast(typing.Any, array)
+    res = random.choices(array, [x['weight'] for x in untyped_array], k=num)
+    print(f' -> picked {res}', file=sys.stderr)
+    return res
+
+
+def pause_exit_spot(pause_evt: threading.Event, resume_evt: threading.Event,
+                    exit_evt: threading.Event) -> bool:
+    """
+    Poll the events, returning True if `exit_evt` was set and pausing as requested by `pause_evt`
+    and `resume_evt`
+    """
+
+    resume_evt.clear()
+    if pause_evt.is_set():
```

Comment:
This doesn’t work though. Consider the following:

    Thread A                          Thread B
                                      # handles /pause
                                      resume_event.clear()
                                      pause_event.set()
    resume_evt.clear()
    if pause_evt.is_set():
                                      # handles /resume
                                      pause_event.clear()
                                      resume_event.set()
        while not resume_evt.wait():
           # ... waits succeeds
                                      # handles /pause
                                      resume_event.clear()
                                      pause_event.set()
        pause_evt.clear()
        ^^^^^^^^^^^^^^^^^
        incorrect pause clear

The clear should not happen in this function.  Only in the HTTP
handler.